docs/architecture/adr-033-protobuf-inter-module-comm.md

```diff
@@ -0,0 +1,339 @@
+any major security threats assuming basic precautions are taken. The basic precaution that the router system will
+need to take is making sure that the `sdk.Context` passed to query methods does not allow writing to the store. This
+can be done for now with a `CacheMultiStore` as is currently done for `BaseApp` queries.
+
+### Internal Methods
+
+In many cases, we may wish for modules to call methods on other modules which are not exposed to clients at all. For this
+purpose, we add the `InternalServer` method to `Configurator`:
+
+```go
+type Configurator interface {
+   MsgServer() grpc.Server
+   QueryServer() grpc.Server
+   InternalServer() grpc.Server
+}
+```
+
+Services registered against `InternalServer` will be callable from other modules but not by external clients.
+
+### Authorization
+
+By default, the inter-module router requires that messages are sent by the first signer returned by `GetSigners`. The
+inter-module router should also accept authorization middleware such as that provided by [ADR 030](https://github.com/cosmos/cosmos-sdk/pull/7105).
+This middleware will allow accounts to otherwise specific module accounts to perform actions on their behalf.
+Authorization middleware should take into account the need to grant certain modules effectively "admin" privileges to
+other modules. This will be addressed in separate ADRs or updates to this ADR.
+
+### Future Work
+
+Other future improvements may include:
+* custom code generation that:
+  * simplifies interfaces (ex. generates code with `sdk.Context` instead of `context.Context`)
+  * optimizes inter-module calls - for instance caching resolved methods after first invocation
+* combining `StoreKey`s and `ModuleKey`s into a single interface so that modules have a single Ocaps handle
+* code generation which makes inter-module communication more performant
+* decoupling `ModuleKey` creation from `AppModuleBasic.Name()` so that app's can override root module account names
+* inter-module hooks and plugins
+
+## Alternatives
+
+The `x/capability` module does provide a proper object-capability implementation that can be used by any module in the
+SDK and could even be used for inter-module Ocaps as described in [\#5931](https://github.com/cosmos/cosmos-sdk/issues/5931).
+
+The advantages of the approach described in this ADR are mostly around how it integrates with other parts of the SDK,
+specifically:
+
+* protobuf so that:
+  * code generation of interfaces can be leveraged for a better dev UX
+  * module interfaces are versioned and checked for breakage using [buf](https://docs.buf.build/breaking-overview)
+* sub-module accounts as per ADR 028
+* the general `Msg` passing paradigm and the way signers are specified by `GetSigners`
+
+Also, this is a complete replacement for keepers and could be applied to _all_ inter-module communication whereas the
+`x/capability` approach in #5931 would need to be applied method by method.
+
```

Comment:
Was a pure go implementation (no proto, not `x/capability`) considered? If so, can it be documented to why it was not chosen?

Reply:
See #7093. I can go into a more detailed comparison if that's helpful

Reply:
oh thanks for this, helps explain the rationale. Can it be documented in the ADR?